### PR TITLE
Add sys_fdatasync

### PR DESCRIPTION
**Pull request by grief8**
This PR adds `fdatasync` support, which is similar to `fsync`, but with a key difference that may improve performance for specific use cases. While `fsync` syncs both the data and metadata of a file to disk, ensuring that both content and attributes (like modification time) are updated on persistent storage, `fdatasync` focuses on flushing only the file's data, not its metadata, unless a change in metadata is crucial for subsequent data retrieval.

## Implementation Details

The implementation of `fdatasync` within this PR closely mirrors that of `fsync`, with the pivotal divergence in their handling of metadata. The provided code adds a `datasync` flag to `sync` and changes the related `sync` implementations.

### CODE REVIEW

File: kernel/aster-nix/src/fs/exfat/inode.rs [hunk 1]
```diff
@@ -568,9 +568,11 @@ impl ExfatInodeInner {
         Ok(())
     }
 
-    fn sync(&self, fs_guard: &MutexGuard<()>) -> Result<()> {
+    fn sync(&self, fs_guard: &MutexGuard<()>, datasync: bool) -> Result<()> {
```
Comment (liqinggd) [inline kernel/aster-nix/src/fs/exfat/inode.rs:571]: `datasync` is not a good name for bool.

File: kernel/aster-nix/src/syscall/fsync.rs [hunk 2]
```diff
@@ -18,6 +18,22 @@ pub fn sys_fsync(fd: FileDesc) -> Result<SyscallReturn> {
             .ok_or(Error::with_message(Errno::EINVAL, "not inode"))?;
         inode_handle.dentry().clone()
     };
-    dentry.sync()?;
+    dentry.sync(false)?;
```
Comment (liqinggd) [inline kernel/aster-nix/src/syscall/fsync.rs:21]: > As well as flushing the file data, fsync() also flushes the metadata information associated with the file.

So fsync should sync the data and metadata. I think you should add another method `sync_data`, which is similar to https://doc.rust-lang.org/std/fs/struct.File.html#method.sync_data. The `sync` method can be renamed to `sync_all`

Comment (grief8) [reply kernel/aster-nix/src/syscall/fsync.rs:21]: Thanks! I think it is better.

**Comment by liqinggd** [comment]
> fdatasync() is similar to fsync(), but does not flush modified
       metadata unless that metadata is needed in order to allow a
       subsequent data retrieval to be correctly handled.  For example,
       changes to st_atime or st_mtime (respectively, time of last
       access and time of last modification; see [inode(7)](https://man7.org/linux/man-pages/man7/inode.7.html)) do not
       require flushing because they are not necessary for a subsequent
       data read to be handled correctly.  On the other hand, a change
       to the file size (st_size, as made by say [ftruncate(2)](https://man7.org/linux/man-pages/man2/ftruncate.2.html)), would
       require a metadata flush.

According to this, the file size should also be flushed if the file size is changed and we call `fdatasync`. It need to refine the implementation of Ext2 and Exfat to group the file_size and data together. I think it is another issue.

**Comment by liqinggd** [comment]
LGTM

**Comment by grief8** [comment]
Hi @tatetian ! Could you please take another look at this PR when you have a chance? Thank you!

File: regression/apps/fdatasync/fdatasync.c [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#include <stdlib.h>
+
+void test_fdatasync_on_fs(const char *directory)
+{
+	char filepath[256];
+	snprintf(filepath, sizeof(filepath), "%s/test_fdatasync.txt",
+		 directory);
+
+	int fd =
+		open(filepath, O_WRONLY | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
```
Comment (tatetian) [inline regression/apps/fdatasync/fdatasync.c:16]: Combine the two lines into one

File: regression/apps/scripts/fs.sh [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    done
+
+    # Clean up
+    rm -f ${test_file}
+    sync
+    cd -
+}
+
+test_fdatasync() {
+    fdatasync/fdatasync /
```
Comment (tatetian) [inline regression/apps/scripts/fs.sh:48]: Are there any sync-related tests from the gvisor test suite?

File: kernel/aster-nix/src/fs/utils/inode.rs [hunk 5]
```diff
@@ -330,7 +330,11 @@ pub trait Inode: Any + Sync + Send {
         Err(Error::new(Errno::EISDIR))
     }
 
-    fn sync(&self) -> Result<()> {
+    fn sync_all(&self) -> Result<()> {
+        Ok(())
+    }
+
+    fn sync_data(&self) -> Result<()> {
```
Comment (tatetian) [inline kernel/aster-nix/src/fs/utils/inode.rs:337]: These two empty implementations can be removed since the trait has provided default implementations.

Comment (grief8) [reply kernel/aster-nix/src/fs/utils/inode.rs:337]: That's right. Thank you for your reminder!

File: regression/apps/scripts/fs.sh (hunk 4 above)
Comment (grief8) [reply regression/apps/scripts/fs.sh:48]: `fsync` is included in the gvisor test suite, and it works well in Asterinas, but `fdatasync` is not covered. Therefore, I just add a simple `fdatasync` test to ensure the functionality.

File: regression/apps/fdatasync/fdatasync.c (hunk 3 above)
Comment (grief8) [reply regression/apps/fdatasync/fdatasync.c:16]: I tried it before. However, it's strange that our format tool enforces it to two separate lines.

**Pull request merged**